include/pybind11/detail/smart_holder_type_casters.h

```diff
@@ -464,6 +464,25 @@ struct shared_ptr_trampoline_self_life_support {
     }
 };
 
+template <typename T,
+          typename D,
+          typename std::enable_if<std::is_default_constructible<D>::value, int>::type = 0>
+inline std::unique_ptr<T, D> unique_with_deleter(T *raw_ptr, std::unique_ptr<D> &&deleter) {
+    if (deleter != nullptr) {
+        return std::unique_ptr<T, D>(raw_ptr, std::move(*deleter));
+    } else {
+        return std::unique_ptr<T, D>(raw_ptr);
+    }
+}
+
+template <typename T,
+          typename D,
+          typename std::enable_if<!std::is_default_constructible<D>::value, int>::type = 0>
+inline std::unique_ptr<T, D> unique_with_deleter(T *raw_ptr, std::unique_ptr<D> &&deleter) {
+    assert(deleter != nullptr);
```

Comment:
@rwgk 
> How troublesome is this to implement: Return nullptr if D is default-constructible, or throw an exception otherwise?

Currently it asserts instead of throwing, this is what I had already written before you suggested this, as I needed it for the actual pointer creation (the `auto result =` line), not sure if you have a strong preference for an assert of throw.

Reply:
I'd definitely put a `pybind11_fail()` here. Otherwise if someone (initially) only tests with `NDEBUG` defined they may see weird behavior. (I know we have some really large binaries that only build with default optimization options.)

Reply:
Done with 4cdd8bbb65bf1e964106e78428095fad80682509.